lib/WUI/nhttp/send_json.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include "step.h"
+
+#include <segmented_json.h>
+#include <http/types.h>
+#include <transfers/monitor.hpp>
+
+#include <string_view>
+
+namespace nhttp::handler {
+
+class TransferState {
+public:
+    std::optional<transfers::TransferId> transfer_id = std::nullopt;
+};
+
+class TransferRenderer final : public json::JsonRenderer<TransferState> {
+public:
+    TransferRenderer()
+        : JsonRenderer(TransferState()) {}
+    virtual json::JsonResult renderState(size_t resume_point, json::JsonOutput &output, TransferState &state) const override;
+};
+
+class Empty {};
+
+/// Renderer with no state attached, will generate json
+/// provided a parameter-less function (not a
+/// closure/lambda or functor) that generates the content. Meant for simple
+/// things where there's no state to track during the generation.
+class EmptyRenderer final : public json::JsonRenderer<Empty> {
+public:
+    typedef json::JsonResult (*Content)(size_t resume_point, json::JsonOutput &output);
+
+private:
+    Content generator;
+
+public:
+    EmptyRenderer(Content content)
+        : JsonRenderer(Empty {})
+        , generator(content) {}
+    virtual json::JsonResult renderState(size_t resume_point, json::JsonOutput &output, Empty &state) const override {
+        return generator(resume_point, output);
+    }
```

Comment:
I wonder if the combination of `virtual` and (implicit) `inline` is a good idea. I mean, the compiler will make _something_ of it, but I wonder what the implications for code size, etc, are.

There's likely no chance for the compiler to opmitize the dynamic dispatch out here anyway, so maybe this could better go into the cpp file?